--- configure.orig	2010-10-04 21:40:28.000000000 +0200
+++ configure	2010-10-04 21:41:40.000000000 +0200
@@ -5135,7 +5135,7 @@
     then
       convert_version_major=`expr "$convert_version" : '.* \([0-9]*\)\.[0-9]*'`
       convert_version_minor=`expr "$convert_version" : '.* [0-9]*\.\([0-9]*\)'`
-      if test "$convert_version_major" -eq 6 -a "$convert_version_minor" -lt 6
+      if test "$convert_version_major" -eq 6 -a "$convert_version_minor" -lt 5
       then
         CONVERT=false
       fi
@@ -5144,7 +5144,7 @@
     then
       { $as_echo "$as_me:${as_lineno-$LINENO}: result: no ($convert_version_major.$convert_version_minor)" >&5
 $as_echo "no ($convert_version_major.$convert_version_minor)" >&6; }
-      as_fn_error $? "You need imagemagick version 6.6 or newer to rebuild icons in maintainer mode." "$LINENO" 5
+      as_fn_error $? "You need imagemagick version 6.5 or newer to rebuild icons in maintainer mode." "$LINENO" 5
     else
       { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes ($convert_version_major.$convert_version_minor)" >&5
 $as_echo "yes ($convert_version_major.$convert_version_minor)" >&6; }
--- configure.ac.orig	2010-10-04 21:42:45.000000000 +0200
+++ configure.ac	2010-10-04 21:43:18.000000000 +0200
@@ -309,7 +309,7 @@
     then
       convert_version_major=`expr "$convert_version" : '.* \([[0-9]]*\)\.[[0-9]]*'`
       convert_version_minor=`expr "$convert_version" : '.* [[0-9]]*\.\([[0-9]]*\)'`
-      if test "$convert_version_major" -eq 6 -a "$convert_version_minor" -lt 6
+      if test "$convert_version_major" -eq 6 -a "$convert_version_minor" -lt 5
       then
         CONVERT=false
       fi
@@ -317,7 +317,7 @@
     if test "$CONVERT" = false
     then
       AC_MSG_RESULT([no ($convert_version_major.$convert_version_minor)])
-      AC_MSG_ERROR([You need imagemagick version 6.6 or newer to rebuild icons in maintainer mode.])
+      AC_MSG_ERROR([You need imagemagick version 6.5 or newer to rebuild icons in maintainer mode.])
     else
       AC_MSG_RESULT([yes ($convert_version_major.$convert_version_minor)])
     fi
