From fa65c91345e0a6f5f47c69c8c1ec7fa370fd42b3 Mon Sep 17 00:00:00 2001
From: Benjamin Drung <bdrung@debian.org>
Date: Fri, 2 Aug 2013 13:04:15 +0200
Subject: [PATCH] Fix D-Bus path for freedesktop.org screen saver.

---
 modules/misc/inhibit/dbus.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/modules/misc/inhibit/dbus.c b/modules/misc/inhibit/dbus.c
index 7106e91..8dc5153 100644
--- a/modules/misc/inhibit/dbus.c
+++ b/modules/misc/inhibit/dbus.c
@@ -57,7 +57,7 @@ static const char dbus_service[][40] =
 
 static const char dbus_path[][33] =
 {
-    [FDO_SS] = "/ScreenSaver",
+    [FDO_SS] = "/org/freedesktop/ScreenSaver",
     [FDO_PM] = "/org/freedesktop/PowerManagement",
     [MATE]   = "/org/mate/SessionManager",
     [GNOME]  = "/org/gnome/SessionManager",
-- 
1.8.1.2

