=== modified file 'modules/misc/inhibit/dbus.c'
--- modules/misc/inhibit/dbus.c	2013-05-27 14:02:21 +0000
+++ modules/misc/inhibit/dbus.c	2014-02-04 17:31:03 +0000
@@ -38,37 +38,31 @@
 
 enum vlc_inhibit_api
 {
-    FDO_SS, /**< KDE >= 4 and GNOME >= 3.6 */
-    FDO_PM, /**< KDE and GNOME <= 2.26 */
-    MATE,  /**< >= 1.0 */
-    GNOME, /**< GNOME 2.26..3.4 */
-};
-
-#define MAX_API (GNOME+1)
-
-/* Currently, all services have identical service and interface names. */
-static const char dbus_service[][40] =
-{
+    FDO_SS, /* KDE, GNOME, Unity */
+    FDO_PM, /* XFCE, LXDE */
+    MATE,
+};
+
+#define MAX_API (MATE+1)
+
+static const char *dbus_service[] =
+{
+    [FDO_SS] = "org.freedesktop.ScreenSaver",
+    [FDO_PM] = "org.freedesktop.PowerManagement",
+    [MATE]   = "org.mate.ScreenSaver",
+};
+
+static const char *dbus_path[] =
+{
+    [FDO_SS] = "/ScreenSaver",
+    [FDO_PM] = "/org/freedesktop/PowerManagement/Inhibit",
+    [MATE]   = "/",
+};
+
+static const char *dbus_interface[] = {
     [FDO_SS] = "org.freedesktop.ScreenSaver",
     [FDO_PM] = "org.freedesktop.PowerManagement.Inhibit",
-    [MATE]   = "org.mate.SessionManager",
-    [GNOME]  = "org.gnome.SessionManager",
-};
-
-static const char dbus_path[][33] =
-{
-    [FDO_SS] = "/ScreenSaver",
-    [FDO_PM] = "/org/freedesktop/PowerManagement",
-    [MATE]   = "/org/mate/SessionManager",
-    [GNOME]  = "/org/gnome/SessionManager",
-};
-
-static const char dbus_method_uninhibit[][10] =
-{
-    [FDO_SS] = "UnInhibit",
-    [FDO_PM] = "UnInhibit",
-    [MATE]   = "Uninhibit",
-    [GNOME]  = "Uninhibit",
+    [MATE]   = "org.mate.ScreenSaver",
 };
 
 struct vlc_inhibit_sys
@@ -115,11 +109,11 @@
         return; /* nothing to do */
 
     /* Send request */
-    const char *method = flags ? "Inhibit" : dbus_method_uninhibit[type];
+    const char *method = flags ? "Inhibit" : "UnInhibit";
     dbus_bool_t ret;
 
     DBusMessage *msg = dbus_message_new_method_call(dbus_service[type],
-                                  dbus_path[type], dbus_service[type], method);
+                                  dbus_path[type], dbus_interface[type], method);
     if (unlikely(msg == NULL))
         return;
 
@@ -130,27 +124,9 @@
 
         assert(sys->cookie == 0);
 
-        switch (type)
-        {
-            case MATE:
-            case GNOME:
-            {
-                dbus_uint32_t xid = 0; // FIXME ?
-                dbus_uint32_t gflags = 0xC;
-
-                ret = dbus_message_append_args(msg, DBUS_TYPE_STRING, &app,
-                                                    DBUS_TYPE_UINT32, &xid,
-                                                    DBUS_TYPE_STRING, &reason,
-                                                    DBUS_TYPE_UINT32, &gflags,
-                                                    DBUS_TYPE_INVALID);
-                break;
-            }
-            default:
-                ret = dbus_message_append_args(msg, DBUS_TYPE_STRING, &app,
-                                                    DBUS_TYPE_STRING, &reason,
-                                                    DBUS_TYPE_INVALID);
-                break;
-        }
+        ret = dbus_message_append_args(msg, DBUS_TYPE_STRING, &app,
+                                            DBUS_TYPE_STRING, &reason,
+                                            DBUS_TYPE_INVALID);
 
         if (!ret
         || !dbus_connection_send_with_reply(sys->conn, msg, &sys->pending, -1))

