diff -x .git -urN linux-3.8/firmware/LICENSE.radeon_rlc zen/firmware/LICENSE.radeon_rlc
--- linux-3.8/firmware/LICENSE.radeon_rlc	1970-01-01 03:00:00.000000000 +0300
+++ zen/firmware/LICENSE.radeon_rlc	2013-04-01 12:17:23.970310827 +0300
@@ -0,0 +1,51 @@
+Copyright (C) 2009, 2010  Advanced Micro Devices, Inc. All rights reserved.
+
+REDISTRIBUTION: Permission is hereby granted, free of any license fees,
+to any person obtaining a copy of this microcode (the "Software"), to
+install, reproduce, copy and distribute copies, in binary form only, of
+the Software and to permit persons to whom the Software is provided to
+do the same, provided that the following conditions are met:
+
+No reverse engineering, decompilation, or disassembly of this Software
+is permitted.
+
+Redistributions must reproduce the above copyright notice, this
+permission notice, and the following disclaimers and notices in the
+Software documentation and/or other materials provided with the
+Software.
+
+DISCLAIMER: THE USE OF THE SOFTWARE IS AT YOUR SOLE RISK.  THE SOFTWARE
+IS PROVIDED "AS IS" AND WITHOUT WARRANTY OF ANY KIND AND COPYRIGHT
+HOLDER AND ITS LICENSORS EXPRESSLY DISCLAIM ALL WARRANTIES, EXPRESS AND
+IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT.
+COPYRIGHT HOLDER AND ITS LICENSORS DO NOT WARRANT THAT THE SOFTWARE WILL
+MEET YOUR REQUIREMENTS, OR THAT THE OPERATION OF THE SOFTWARE WILL BE
+UNINTERRUPTED OR ERROR-FREE.  THE ENTIRE RISK ASSOCIATED WITH THE USE OF
+THE SOFTWARE IS ASSUMED BY YOU.  FURTHERMORE, COPYRIGHT HOLDER AND ITS
+LICENSORS DO NOT WARRANT OR MAKE ANY REPRESENTATIONS REGARDING THE USE
+OR THE RESULTS OF THE USE OF THE SOFTWARE IN TERMS OF ITS CORRECTNESS,
+ACCURACY, RELIABILITY, CURRENTNESS, OR OTHERWISE.
+
+DISCLAIMER: UNDER NO CIRCUMSTANCES INCLUDING NEGLIGENCE, SHALL COPYRIGHT
+HOLDER AND ITS LICENSORS OR ITS DIRECTORS, OFFICERS, EMPLOYEES OR AGENTS
+("AUTHORIZED REPRESENTATIVES") BE LIABLE FOR ANY INCIDENTAL, INDIRECT,
+SPECIAL OR CONSEQUENTIAL DAMAGES (INCLUDING DAMAGES FOR LOSS OF BUSINESS
+PROFITS, BUSINESS INTERRUPTION, LOSS OF BUSINESS INFORMATION, AND THE
+LIKE) ARISING OUT OF THE USE, MISUSE OR INABILITY TO USE THE SOFTWARE,
+BREACH OR DEFAULT, INCLUDING THOSE ARISING FROM INFRINGEMENT OR ALLEGED
+INFRINGEMENT OF ANY PATENT, TRADEMARK, COPYRIGHT OR OTHER INTELLECTUAL
+PROPERTY RIGHT EVEN IF COPYRIGHT HOLDER AND ITS AUTHORIZED
+REPRESENTATIVES HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.  IN
+NO EVENT SHALL COPYRIGHT HOLDER OR ITS AUTHORIZED REPRESENTATIVES TOTAL
+LIABILITY FOR ALL DAMAGES, LOSSES, AND CAUSES OF ACTION (WHETHER IN
+CONTRACT, TORT (INCLUDING NEGLIGENCE) OR OTHERWISE) EXCEED THE AMOUNT OF
+US$10.
+
+Notice:  The Software is subject to United States export laws and
+regulations.  You agree to comply with all domestic and international
+export laws and regulations that apply to the Software, including but
+not limited to the Export Administration Regulations administered by the
+U.S. Department of Commerce and International Traffic in Arm Regulations
+administered by the U.S. Department of State.  These laws include
+restrictions on destinations, end users and end use.
Binary files linux-3.8/firmware/radeon/CEDAR_me.bin and zen/firmware/radeon/CEDAR_me.bin differ
Binary files linux-3.8/firmware/radeon/CEDAR_pfp.bin and zen/firmware/radeon/CEDAR_pfp.bin differ
Binary files linux-3.8/firmware/radeon/CEDAR_rlc.bin and zen/firmware/radeon/CEDAR_rlc.bin differ
Binary files linux-3.8/firmware/radeon/CYPRESS_me.bin and zen/firmware/radeon/CYPRESS_me.bin differ
Binary files linux-3.8/firmware/radeon/CYPRESS_pfp.bin and zen/firmware/radeon/CYPRESS_pfp.bin differ
Binary files linux-3.8/firmware/radeon/CYPRESS_rlc.bin and zen/firmware/radeon/CYPRESS_rlc.bin differ
Binary files linux-3.8/firmware/radeon/JUNIPER_me.bin and zen/firmware/radeon/JUNIPER_me.bin differ
Binary files linux-3.8/firmware/radeon/JUNIPER_pfp.bin and zen/firmware/radeon/JUNIPER_pfp.bin differ
Binary files linux-3.8/firmware/radeon/JUNIPER_rlc.bin and zen/firmware/radeon/JUNIPER_rlc.bin differ
Binary files linux-3.8/firmware/radeon/R600_rlc.bin and zen/firmware/radeon/R600_rlc.bin differ
Binary files linux-3.8/firmware/radeon/R700_rlc.bin and zen/firmware/radeon/R700_rlc.bin differ
Binary files linux-3.8/firmware/radeon/REDWOOD_me.bin and zen/firmware/radeon/REDWOOD_me.bin differ
Binary files linux-3.8/firmware/radeon/REDWOOD_pfp.bin and zen/firmware/radeon/REDWOOD_pfp.bin differ
Binary files linux-3.8/firmware/radeon/REDWOOD_rlc.bin and zen/firmware/radeon/REDWOOD_rlc.bin differ
diff -x .git -urN linux-3.8/firmware/WHENCE zen/firmware/WHENCE
--- linux-3.8/firmware/WHENCE	2013-02-19 01:58:34.000000000 +0200
+++ zen/firmware/WHENCE	2013-04-01 12:17:24.014310277 +0300
@@ -862,3 +862,24 @@
 Found in hex form in kernel source.
 
 --------------------------------------------------------------------------
+
+Driver: radeon - ATI Radeon
+
+File: radeon/R600_rlc.bin
+File: radeon/R700_rlc.bin
+File: radeon/CEDAR_me.bin
+File: radeon/CEDAR_pfp.bin
+File: radeon/CEDAR_rlc.bin
+File: radeon/CYPRESS_me.bin
+File: radeon/CYPRESS_pfp.bin
+File: radeon/CYPRESS_rlc.bin
+File: radeon/JUNIPER_me.bin
+File: radeon/JUNIPER_pfp.bin
+File: radeon/JUNIPER_rlc.bin
+File: radeon/REDWOOD_me.bin
+File: radeon/REDWOOD_pfp.bin
+File: radeon/REDWOOD_rlc.bin
+
+Licence: Redistributable. See LICENSE.radeon_rlc for details.
+
+--------------------------------------------------------------------------
