diff -x .git -urN linux-3.8/net/ipv4/Kconfig zen/net/ipv4/Kconfig
--- linux-3.8/net/ipv4/Kconfig	2013-02-19 01:58:34.000000000 +0200
+++ zen/net/ipv4/Kconfig	2013-02-19 12:18:55.193643514 +0200
@@ -600,6 +600,9 @@
 	config DEFAULT_VEGAS
 		bool "Vegas" if TCP_CONG_VEGAS=y
 
+	config DEFAULT_YEAH
+		bool "YeAH" if TCP_CONG_YEAH=y
+
 	config DEFAULT_VENO
 		bool "Veno" if TCP_CONG_VENO=y
 
@@ -625,6 +628,7 @@
 	default "htcp" if DEFAULT_HTCP
 	default "hybla" if DEFAULT_HYBLA
 	default "vegas" if DEFAULT_VEGAS
+	default "yeah" if DEFAULT_YEAH
 	default "westwood" if DEFAULT_WESTWOOD
 	default "veno" if DEFAULT_VENO
 	default "reno" if DEFAULT_RENO
