diff -urN old/vmblock-only/Makefile.kernel work/vmblock-only/Makefile.kernel
--- old/vmblock-only/Makefile.kernel	2009-08-21 16:10:32.000000000 -0400
+++ work/vmblock-only/Makefile.kernel	2009-08-21 16:11:34.000000000 -0400
@@ -19,7 +19,7 @@
 
 INCLUDE := -I$(SRCROOT)/include
 
-EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE)
+EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE) $(LINUXINCLUDE)
 
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/setnice.c, -DVMW_HAVE_SET_USER_NICE, )
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/epoll.c,   -DVMW_HAVE_EPOLL, )
diff -urN old/vmci-only/Makefile.kernel work/vmci-only/Makefile.kernel
--- old/vmci-only/Makefile.kernel	2009-08-21 16:10:32.000000000 -0400
+++ work/vmci-only/Makefile.kernel	2009-08-21 16:11:47.000000000 -0400
@@ -21,7 +21,7 @@
 
 INCLUDE := -I$(SRCROOT)/include -I$(SRCROOT)/common -I$(SRCROOT)/linux
 
-EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE)
+EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE) $(LINUXINCLUDE)
 
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/epoll.c,   -DVMW_HAVE_EPOLL, )
 
diff -urN old/vmmon-only/Makefile.kernel work/vmmon-only/Makefile.kernel
--- old/vmmon-only/Makefile.kernel	2009-08-21 16:10:32.000000000 -0400
+++ work/vmmon-only/Makefile.kernel	2009-08-21 16:12:48.000000000 -0400
@@ -22,7 +22,7 @@
 INCLUDE := -I$(SRCROOT)/include -I$(SRCROOT)/common -I$(SRCROOT)/linux \
 	   -I$(SRCROOT)/vmcore
 
-EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE)
+EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE) $(LINUXINCLUDE)
 
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/nopage1.c, -DVMW_NOPAGE_261, )
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/skas1.c,   -DVMW_SKAS_MMAP, )
diff -urN old/vmnet-only/Makefile.kernel work/vmnet-only/Makefile.kernel
--- old/vmnet-only/Makefile.kernel	2009-08-21 16:10:32.000000000 -0400
+++ work/vmnet-only/Makefile.kernel	2009-08-21 16:13:02.000000000 -0400
@@ -19,7 +19,7 @@
 
 INCLUDE := -I$(SRCROOT)
 
-EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE)
+EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE) $(LINUXINCLUDE)
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/epoll.c,   -DVMW_HAVE_EPOLL, )
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/sk_alloc.c,-DVMW_HAVE_SK_ALLOC_WITH_PROTO, )
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/netdev_has_net.c,-DVMW_NETDEV_HAS_NET, )
diff -urN old/vsock-only/linux/util.c work/vsock-only/linux/util.c
--- old/vsock-only/linux/util.c	2009-08-21 16:10:32.000000000 -0400
+++ work/vsock-only/linux/util.c	2009-08-21 16:13:27.000000000 -0400
@@ -154,7 +154,7 @@
       goto error;
    }
 
-   Log(buf);
+   Log("%s", buf);
 
    return;
 
diff -urN old/vsock-only/Makefile.kernel work/vsock-only/Makefile.kernel
--- old/vsock-only/Makefile.kernel	2009-08-21 16:10:32.000000000 -0400
+++ work/vsock-only/Makefile.kernel	2009-08-21 16:13:41.000000000 -0400
@@ -25,7 +25,7 @@
 INCLUDE += -I$(SRCROOT)/linux
 INCLUDE += -I$(SRCROOT)/common
 
-EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE)
+EXTRA_CFLAGS := $(CC_OPTS) $(INCLUDE) $(LINUXINCLUDE)
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/epoll.c,   -DVMW_HAVE_EPOLL, )
 EXTRA_CFLAGS += $(call vm_check_build, $(SRCROOT)/autoconf/setnice.c, -DVMW_HAVE_SET_USER_NICE, )
 
