diff -c work/kernel/conftest.sh work-new/kernel//conftest.sh
*** work/kernel/conftest.sh	2011-07-22 20:33:34.713262885 +0300
--- work-new/kernel//conftest.sh	2011-07-22 20:35:18.644963558 +0300
***************
*** 76,82 ****
  }
  
  build_cflags() {
!     BASE_CFLAGS="-D__KERNEL__ \
  -DKBUILD_BASENAME=\"#conftest$$\" -DKBUILD_MODNAME=\"#conftest$$\" \
  -nostdinc -isystem $ISYSTEM"
  
--- 76,82 ----
  }
  
  build_cflags() {
!     BASE_CFLAGS="-O2 -D__KERNEL__ \
  -DKBUILD_BASENAME=\"#conftest$$\" -DKBUILD_MODNAME=\"#conftest$$\" \
  -nostdinc -isystem $ISYSTEM"
  
diff -c work/kernel/nv-linux.h work-new/kernel//nv-linux.h
*** work/kernel/nv-linux.h	2011-05-17 09:32:19.000000000 +0300
--- work-new/kernel//nv-linux.h	2011-07-22 20:35:18.645963546 +0300
***************
*** 32,38 ****
  #  define KERNEL_2_4
  #elif LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
  #  error This driver does not support 2.5 kernels!
! #elif LINUX_VERSION_CODE < KERNEL_VERSION(2, 7, 0)
  #  define KERNEL_2_6
  #else
  #  error This driver does not support development kernels!
--- 32,38 ----
  #  define KERNEL_2_4
  #elif LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
  #  error This driver does not support 2.5 kernels!
! #elif LINUX_VERSION_CODE < KERNEL_VERSION(3, 1, 0)
  #  define KERNEL_2_6
  #else
  #  error This driver does not support development kernels!
