commit 27053f1a7ebac04d56e282f21e7e0d10187ba1a7
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Tue Jan 31 11:17:12 2012 +0400

    Add update branch for overlays.

diff --git a/pym/cl_builder.py b/pym/cl_builder.py
index 34cd1d5..583daa7 100644
--- a/pym/cl_builder.py
+++ b/pym/cl_builder.py
@@ -265,7 +265,8 @@ class cl_builder(color_print):
             for gitName,gitDir in (('portage','/usr/portage'),
                            ('overlay','/var/lib/layman/calculate')):
                 gitHead = pathJoin(builderPath,gitDir,".git/HEAD")
-                if "heads/develop" in readFile(gitHead):
+                if "heads/develop" in readFile(gitHead) or \
+                   "heads/update" in readFile(gitHead):
                     self.printMessageForTest(
                         _("Changing {repname} repository branch to '{branch}'").
                             format(repname=gitName,
