diff --git pym/cl_builder.py pym/cl_builder.py
index 8070896..aac5474 100644
--- pym/cl_builder.py
+++ pym/cl_builder.py
@@ -140,7 +140,7 @@ class cl_builder(color_print):
         #self.clVars.Set("cl_root_path","/", True)
         self.clVars.Set("cl_action","iso", True)
         chrootPath=pathJoin(directory,
-                            self.clVars.Get('cl_builder_squash_path'))
+            self.clVars.Get('cl_builder_squash_path'))
         self.clVars.Set("cl_chroot_path",chrootPath, True)
         self.clVars.Set("cl_root_path","..", True)
         self.clTempl = template(self.clVars)
@@ -254,6 +254,8 @@ class cl_builder(color_print):
                 raise BuilderError(_("Failed to unmount %s")%target)
             self.printByResult(True)
         distrPath = distr.convertToDirectory().getDirectory()
+        self.clVars.Set('cl_builder_squash_path',
+                        os.path.basename(os.path.normpath(distrPath)))
         self.cleanNeedlessKernelData(distrPath)
         self.dispatchConf(distrPath)
 
