diff --git pym/cl_builder.py pym/cl_builder.py
index 7a72d81..d4078a2 100644
--- pym/cl_builder.py
+++ pym/cl_builder.py
@@ -666,9 +666,7 @@ class cl_builder(color_print):
             return False
         self.printSUCCESS(_("System built successfully"))
         if self.assembleIso:
-            self.printWARNING(_("If you want to emerge packages manually "
-                        "on the system you are assembling, please execute:"))
-            self.printWARNING("  cl-make -p %s -T"%
+            os.system("cl-make -p %s -fT"%
                               self.clVars.Get('os_builder_profile'))
         return True
 
