--- a/deadbeef.desktop.in.org	2019-04-27 22:19:36.685464263 -0000
+++ b/deadbeef.desktop.in	2019-04-27 22:22:48.415836923 -0000
@@ -20,32 +20,4 @@
 Keywords=Sound;Music;Audio;Player;Musicplayer;MP3;
 Keywords[zh_TW]=Sound;Music;Audio;Player;Musicplayer;MP3;音樂;音樂播放器;播放器;音訊;
 
-X-Ayatana-Desktop-Shortcuts=Play;Pause;Stop;Next;Prev;
 X-PulseAudio-Properties=media.role=music
-
-[Desktop Action Play]
-Name=Play
-Name[zh_CN]=播放
-Name[zh_TW]=播放
-Exec=deadbeef --play
-
-[Desktop Action Pause]
-Name=Pause
-Name[zh_CN]=暂停
-Name[zh_TW]=暫停
-Exec=deadbeef --pause
-
-[Desktop Action Stop]
-Name=Stop
-Name[zh_TW]=停止
-Exec=deadbeef --stop
-
-[Desktop Action Next]
-Name=Next
-Name[zh_TW]=下一首
-Exec=deadbeef --next
-
-[Desktop Action Prev]
-Name=Prev
-Name[zh_TW]=上一首
-Exec=deadbeef --prev
