--- vmblock-only/autoconf/cachecreate.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/cachecreate.c.new	2010-02-05 23:29:47.077574422 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 /*
  * All kernels before 2.6.22 take 6 arguments.  All kernels since
--- vmblock-only/autoconf/cachector.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/cachector.c.new	2010-02-05 23:32:39.086240477 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 /*
  * Between 2.6.23 and 2.6.24-rc1 ctor prototype was changed from
--- vmblock-only/autoconf/cachector1.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/cachector1.c.new	2010-02-05 23:33:57.863050887 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 /*
  * Between 2.6.27-rc1 and 2.6.27-rc2 ctor prototype was changed from
--- vmblock-only/autoconf/epoll.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/epoll.c.new	2010-02-05 23:34:03.782195888 +0100
@@ -22,8 +22,12 @@
  * never had it, but some distros backported epoll patch.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
 #include <linux/poll.h>
--- vmblock-only/autoconf/filldir1.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/filldir1.c.new	2010-02-05 23:34:09.582480900 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 20)
 #include <linux/fs.h>
--- vmblock-only/autoconf/geninclude.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/geninclude.c.new	2010-02-05 23:34:16.775702411 +0100
@@ -16,7 +16,12 @@
  *
  *********************************************************/
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #ifdef CONFIG_X86_VOYAGER
 APATH/mach-voyager
--- vmblock-only/autoconf/getsb1.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/getsb1.c.new	2010-02-05 23:34:21.727106345 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 
--- vmblock-only/autoconf/setnice.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/setnice.c.new	2010-02-05 23:34:27.326719885 +0100
@@ -20,8 +20,12 @@
  * set_user_nice appeared in 2.4.21.  But some distros
  * backported it to older kernels.
  */
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 21)
 #include <linux/sched.h>
--- vmblock-only/autoconf/statfs1.c	2009-03-27 06:34:29.000000000 +0100
+++ vmblock-only/autoconf/statfs1.c.new	2010-02-05 23:34:34.998496715 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 #include <linux/fs.h>
--- vmblock-only/include/driver-config.h.old	2010-02-03 00:12:31.439000354 +0100
+++ vmblock-only/include/driver-config.h	2010-02-03 00:13:29.608823892 +0100
@@ -32,7 +32,12 @@
 #define INCLUDE_ALLOW_VMNIXMOD
 #include "includeCheck.h"

+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 #include "compat_version.h"
 
 /*
--- vmci-only/autoconf/epoll.c	2009-03-27 06:34:31.000000000 +0100
+++ vmci-only/autoconf/epoll.c.new	2010-02-05 23:38:31.793973380 +0100
@@ -22,8 +22,12 @@
  * never had it, but some distros backported epoll patch.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
 #include <linux/poll.h>
--- vmci-only/autoconf/geninclude.c	2009-03-27 06:34:31.000000000 +0100
+++ vmci-only/autoconf/geninclude.c.new	2010-02-05 23:38:45.704327021 +0100
@@ -16,7 +16,12 @@
  *
  *********************************************************/
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #ifdef CONFIG_X86_VOYAGER
 APATH/mach-voyager
--- vmci-only/include/driver-config.h.old	2010-02-03 00:12:31.439000354 +0100
+++ vmci-only/include/driver-config.h	2010-02-03 00:13:29.608823892 +0100
@@ -32,7 +32,12 @@
 #define INCLUDE_ALLOW_VMNIXMOD
 #include "includeCheck.h"
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 #include "compat_version.h"
 
 /*
--- vmmon-only/autoconf/epoll.c	2010-02-05 23:39:45.361576611 +0100
+++ vmmon-only/autoconf/epoll.c.new	2010-02-05 23:40:41.241245910 +0100
@@ -22,8 +22,12 @@
  * never had it, but some distros backported epoll patch.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
 #include <linux/poll.h>
--- vmmon-only/autoconf/geninclude.c	2010-02-05 23:39:45.362577852 +0100
+++ vmmon-only/autoconf/geninclude.c.new	2010-02-05 23:40:50.833402279 +0100
@@ -16,7 +16,12 @@
  *
  *********************************************************/
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #ifdef CONFIG_X86_VOYAGER
 APATH/mach-voyager
--- vmmon-only/autoconf/nopage1.c	2010-02-05 23:39:45.363577625 +0100
+++ vmmon-only/autoconf/nopage1.c.new	2010-02-05 23:40:58.674161808 +0100
@@ -21,8 +21,12 @@
  * it uses new prototype.  For 2.6.1 (and its -rc) and older do
  * compile test.
  */
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 2)
 #include <linux/mm.h>
--- vmmon-only/autoconf/setnice.c	2010-02-05 23:39:45.363577625 +0100
+++ vmmon-only/autoconf/setnice.c.new	2010-02-05 23:41:12.017985901 +0100
@@ -20,8 +20,12 @@
  * set_user_nice appeared in 2.4.21.  But some distros
  * backported it to older kernels.
  */
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 21)
 #include <linux/sched.h>
--- vmmon-only/autoconf/smpcall.c	2010-02-05 23:39:45.364577754 +0100
+++ vmmon-only/autoconf/smpcall.c.new	2010-02-05 23:40:25.057954672 +0100
@@ -21,8 +21,12 @@
  * Change happened between 2.6.26 and 2.6.27-rc1.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 26)
 #   error This compile test intentionally fails.
--- vmmon-only/include/driver-config.h.old	2010-02-03 00:12:31.439000354 +0100
+++ vmmon-only/include/driver-config.h	2010-02-03 00:13:29.608823892 +0100
@@ -32,7 +32,12 @@
 #define INCLUDE_ALLOW_VMNIXMOD
 #include "includeCheck.h"
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 #include "compat_version.h"
 
 /*
--- vmnet-only/epoll.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/epoll.c.new	2010-02-05 23:43:43.642940246 +0100
@@ -22,8 +22,12 @@
  * never had it, but some distros backported epoll patch.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
 #include <linux/poll.h>
--- vmnet-only/geninclude.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/geninclude.c.new	2010-02-05 23:43:52.931081128 +0100
@@ -16,7 +16,12 @@
  *
  *********************************************************/
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #ifdef CONFIG_X86_VOYAGER
 APATH/mach-voyager
--- vmnet-only/netdev_has_dev_net.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/netdev_has_dev_net.c.new	2010-02-05 23:44:11.315557541 +0100
@@ -21,8 +21,12 @@
  * It appeared between 2.6.25 and 2.6.26-rc1.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 25)
 #   error This compile test intentionally fails.
--- vmnet-only/netdev_has_net.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/netdev_has_net.c.new	2010-02-05 23:44:29.348948588 +0100
@@ -27,8 +27,12 @@
  * as hopefully in few weeks all 2.6.24-rc1 users will be gone.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 23)
 #   error This compile test intentionally fails.
--- vmnet-only/nfhook_uses_skb.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/nfhook_uses_skb.c.new	2010-02-05 23:44:35.228702615 +0100
@@ -23,8 +23,12 @@
  * 2.6.24-rc1.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 23)
 #   error This compile test intentionally fails.
--- vmnet-only/setnice.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/setnice.c.new	2010-02-05 23:44:45.613286546 +0100
@@ -20,8 +20,12 @@
  * set_user_nice appeared in 2.4.21.  But some distros
  * backported it to older kernels.
  */
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 21)
 #include <linux/sched.h>
--- vmnet-only/sk_alloc.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/sk_alloc.c.new	2010-02-05 23:44:59.796490045 +0100
@@ -21,8 +21,12 @@
  * This API change was introduced between 2.6.12-rc1 and 2.6.12-rc2.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 13)
 #include <net/sock.h>
--- vmnet-only/skblin.c	2009-03-27 06:34:28.000000000 +0100
+++ vmnet-only/skblin.c.new	2010-02-05 23:45:17.276123591 +0100
@@ -20,8 +20,12 @@
  * Detect whether skb_linearize takes one or two arguments.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 17)
 /*
--- vmnet-only/driver-config.h.old	2010-02-03 00:12:31.439000354 +0100
+++ vmnet-only/driver-config.h	2010-02-03 00:13:29.608823892 +0100
@@ -32,7 +32,12 @@
 #define INCLUDE_ALLOW_VMNIXMOD
 #include "includeCheck.h"
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 #include "compat_version.h"
 
 /*
--- vsock-only/autoconf/cachecreate.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/cachecreate.c.new	2010-02-05 23:46:29.157369452 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 /*
  * All kernels before 2.6.22 take 6 arguments.  All kernels since
--- vsock-only/autoconf/epoll.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/epoll.c.new	2010-02-05 23:46:39.884406108 +0100
@@ -22,8 +22,12 @@
  * never had it, but some distros backported epoll patch.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
 #include <linux/poll.h>
--- vsock-only/autoconf/filldir1.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/filldir1.c.new	2010-02-05 23:46:52.692243523 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 20)
 #include <linux/fs.h>
--- vsock-only/autoconf/geninclude.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/geninclude.c.new	2010-02-05 23:46:58.292953391 +0100
@@ -16,7 +16,12 @@
  *
  *********************************************************/
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #ifdef CONFIG_X86_VOYAGER
 APATH/mach-voyager
--- vsock-only/autoconf/getsb1.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/getsb1.c.new	2010-02-05 23:47:07.516892747 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 
--- vsock-only/autoconf/setnice.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/setnice.c.new	2010-02-05 23:47:16.829514462 +0100
@@ -20,8 +20,12 @@
  * set_user_nice appeared in 2.4.21.  But some distros
  * backported it to older kernels.
  */
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 4, 21)
 #include <linux/sched.h>
--- vsock-only/autoconf/sk_filter.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/sk_filter.c.new	2010-02-05 23:47:28.501825992 +0100
@@ -23,8 +23,12 @@
  * This test will fail to build on kernels with the new interface.
  */
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 /*
  * We'd restrict this test to 2.4.21 and earlier kernels, but Mandrake's
--- vsock-only/autoconf/statfs1.c	2009-03-27 06:34:32.000000000 +0100
+++ vsock-only/autoconf/statfs1.c.new	2010-02-05 23:47:41.174472911 +0100
@@ -16,8 +16,12 @@
  *
  *********************************************************/
 
-#include <linux/autoconf.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
+#include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 #include <linux/fs.h>
--- vsock-only/include/driver-config.h.old	2010-02-03 00:12:31.439000354 +0100
+++ vsock-only/include/driver-config.h	2010-02-03 00:13:29.608823892 +0100
@@ -32,7 +32,12 @@
 #define INCLUDE_ALLOW_VMNIXMOD
 #include "includeCheck.h"
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
 #include <linux/autoconf.h>
+#else
+#include <generated/autoconf.h>
+#endif
 #include "compat_version.h"
 
 /*
