diff --git a/makefile b/makefile
index 282d7ef..c8c4270 100755
--- a/makefile
+++ b/makefile
@@ -22,7 +22,7 @@ build:
 	$(CC) -Isrc/include `pkg-config --cflags gtk+-3.0` $(BUILD_FILES_GUI) -o zenmonitor `pkg-config --libs gtk+-3.0` -lm -no-pie -Wall $(CFLAGS)
 
 build-cli:
-	$(CC) -Isrc/include `pkg-config --cflags glib-2.0` $(BUILD_FILES_CLI) -o zenmonitor-cli `pkg-config --libs glib-2.0` -lm -lncurses -no-pie -Wall $(CFLAGS)
+	$(CC) -Isrc/include `pkg-config --cflags glib-2.0` $(BUILD_FILES_CLI) -o zenmonitor-cli `pkg-config --libs glib-2.0` -lm -lncurses -ltinfo -no-pie -Wall $(CFLAGS)
 
 install:
 	mkdir -p $(DESTDIR)$(PREFIX)/bin
