diff --git a/tools/seafile-admin b/tools/seafile-admin
index e127d55..3eb769a 100755
--- a/tools/seafile-admin
+++ b/tools/seafile-admin
@@ -727,7 +727,7 @@ def read_seafile_data_dir(ccnet_conf_dir):
     seafile_ini = os.path.join(ccnet_conf_dir, 'seafile.ini')
     if os.path.exists(seafile_ini):
         with open(seafile_ini, 'r') as fp:
-            seafile_data_dir = fp.read()
+            seafile_data_dir = fp.read().strip('\n')
     else:
         # In previous seafile-admin, seafiled-data folder must be under
         # the top level directory, so we do not store the location of
@@ -984,4 +984,4 @@ def main():
     args.func(args)
 
 if __name__ == '__main__':
-    main()
\ No newline at end of file
+    main()
