--- a/drivers/net/ethernet/broadcom/bnx2.c	2012-07-03 18:43:03.404285359 -0400
+++ b/drivers/net/ethernet/broadcom/bnx2.c	2012-07-03 18:57:48.459622386 -0400
@@ -59,7 +59,7 @@
 #define DRV_MODULE_NAME		"bnx2"
 #define DRV_MODULE_VERSION	"2.2.1"
 #define DRV_MODULE_RELDATE	"Dec 18, 2011"
-#define FW_MIPS_FILE_06		"bnx2/bnx2-mips-06-6.2.3.fw"
+#define FW_MIPS_FILE_06         "bnx2/bnx2-mips-06-6.2.1.fw"
 #define FW_RV2P_FILE_06		"bnx2/bnx2-rv2p-06-6.0.15.fw"
 #define FW_MIPS_FILE_09		"bnx2/bnx2-mips-09-6.2.1b.fw"
 #define FW_RV2P_FILE_09_Ax	"bnx2/bnx2-rv2p-09ax-6.0.17.fw"
