diff -Nru kernel-2.6.19mamba.orig/sound/pci/hda/hda_generic.c kernel-2.6.19mamba/sound/pci/hda/hda_generic.c
--- kernel-2.6.19mamba.orig/sound/pci/hda/hda_generic.c	2006-11-29 22:57:37.000000000 +0100
+++ kernel-2.6.19mamba/sound/pci/hda/hda_generic.c	2007-01-02 17:32:42.000000000 +0100
@@ -485,7 +485,7 @@
 			return "Front Aux";
 		return "Aux";
 	case AC_JACK_MIC_IN:
-		if (node->pin_caps &
+		if (pinctl && node->pin_caps &
 		    (AC_PINCAP_VREF_80 << AC_PINCAP_VREF_SHIFT))
 			*pinctl |= AC_PINCTL_VREF_80;
 		if ((location & 0x0f) == AC_JACK_LOC_FRONT)
