--- linux-2.6.orig/arch/powerpc/Kconfig 2007-03-19 11:47:27.422413925 +0100
+++ linux-2.6/arch/powerpc/Kconfig      2007-03-19 11:47:37.512413925 +0100
@@ -11,6 +11,11 @@  config PPC64
          This option selects whether a 32-bit or a 64-bit kernel
          will be built.
 
+config PPC_PM_NEEDS_RTC_LIB
+       bool
+       select RTC_LIB
+       default y if PM
+
 config PPC32
        bool
        default y if !PPC64
--- linux-2.6.orig/arch/powerpc/sysdev/Makefile 2007-03-19 11:47:27.532413925 +0100
+++ linux-2.6/arch/powerpc/sysdev/Makefile      2007-03-19 11:47:37.512413925 +0100
@@ -14,6 +14,9 @@  obj-$(CONFIG_FSL_SOC)                += fsl_soc.o
 obj-$(CONFIG_TSI108_BRIDGE)    += tsi108_pci.o tsi108_dev.o
 obj-$(CONFIG_QUICC_ENGINE)     += qe_lib/
 
+# contains only the suspend handler for time
+obj-$(CONFIG_PM)               += timer.o
+
 ifeq ($(CONFIG_PPC_MERGE),y)
 obj-$(CONFIG_PPC_I8259)                += i8259.o
 obj-$(CONFIG_PPC_83xx)         += ipic.o
