--- kernel-2.6.26mamba.orig/kernel/rcupreempt.c 2008-07-13 23:51:29.000000000 +0200
+++ kernel-2.6.26mamba/kernel/rcupreempt.c      2008-09-15 01:42:19.000000000 +0200
@@ -281,7 +281,7 @@
                local_irq_restore(flags);
        }
 }
-EXPORT_SYMBOL_GPL(__rcu_read_lock);
+EXPORT_SYMBOL(__rcu_read_lock);

 void __rcu_read_unlock(void)
 {
@@ -351,7 +351,7 @@
                local_irq_restore(flags);
        }
 }
-EXPORT_SYMBOL_GPL(__rcu_read_unlock);
+EXPORT_SYMBOL(__rcu_read_unlock);

 /*
  * If a global counter flip has occurred since the last time that we

