diff -Nru kernel-2.6.21mamba.orig/drivers/kvm/x86_emulate.c kernel-2.6.21mamba/drivers/kvm/x86_emulate.c
--- kernel-2.6.21mamba.orig/drivers/kvm/x86_emulate.c	2007-04-26 05:08:32.000000000 +0200
+++ kernel-2.6.21mamba/drivers/kvm/x86_emulate.c	2007-06-05 03:40:34.000000000 +0200
@@ -152,7 +152,7 @@
 static u16 twobyte_table[256] = {
 	/* 0x00 - 0x0F */
 	0, SrcMem | ModRM | DstReg, 0, 0, 0, 0, ImplicitOps, 0,
-	0, 0, 0, 0, 0, ImplicitOps | ModRM, 0, 0,
+	0, ImplicitOps, 0, 0, 0, ImplicitOps | ModRM, 0, 0,
 	/* 0x10 - 0x1F */
 	0, 0, 0, 0, 0, 0, 0, 0, ImplicitOps | ModRM, 0, 0, 0, 0, 0, 0, 0,
 	/* 0x20 - 0x2F */
@@ -1303,6 +1303,8 @@
 	/* Disable writeback. */
 	dst.orig_val = dst.val;
 	switch (b) {
+	case 0x09:		/* wbinvd */
+		break;
 	case 0x0d:		/* GrpP (prefetch) */
 	case 0x18:		/* Grp16 (prefetch/nop) */
 		break;
