diff --git a/drivers/clk/clk.c b/drivers/clk/clk.c
index 251e45d..fbe0f3a 100644
--- a/drivers/clk/clk.c
+++ b/drivers/clk/clk.c
@@ -259,10 +259,11 @@  late_initcall(clk_disable_unused);
 
 /***    helper functions   ***/
 
-inline const char *__clk_get_name(struct clk *clk)
+const char *__clk_get_name(struct clk *clk)
 {
 	return !clk ? NULL : clk->name;
 }
+EXPORT_SYMBOL_GPL(__clk_get_name);
 
 inline struct clk_hw *__clk_get_hw(struct clk *clk)
 {
