--- kernel-3.2mamba/drivers/net/usb/asix.c.orig	2012-03-11 17:40:47.570425305 +0100
+++ kernel-3.2mamba/drivers/net/usb/asix.c	2012-03-11 18:02:23.178248012 +0100
@@ -1580,6 +1580,10 @@
 	USB_DEVICE (0x6189, 0x182d),
 	.driver_info =  (unsigned long) &ax8817x_info,
 }, {
+	// Sitecom LN-031 "USB 2.0 10/100/1000 Ethernet adapter"
+	USB_DEVICE (0x0df6, 0x0056),
+	.driver_info =  (unsigned long) &ax88178_info,
+}, {
 	// corega FEther USB2-TX
 	USB_DEVICE (0x07aa, 0x0017),
 	.driver_info =  (unsigned long) &ax8817x_info,
