diff -Nru kernel-2.6.33mamba.orig//drivers/usb/serial/ftdi_sio.c kernel-2.6.33mamba/drivers/usb/serial/ftdi_sio.c
--- kernel-2.6.33mamba.orig//drivers/usb/serial/ftdi_sio.c	2010-09-29 13:08:53.454057663 +0200
+++ kernel-2.6.33mamba/drivers/usb/serial/ftdi_sio.c	2010-09-29 13:18:30.229054171 +0200
@@ -150,6 +150,8 @@
  * /sys/bus/usb/ftdi_sio/new_id, then send patch/report!
  */
 static struct usb_device_id id_table_combined [] = {
+	{ USB_DEVICE(FTDI_VID, FTDI_CTI_Mini_PID) },
+	{ USB_DEVICE(FTDI_VID, FTDI_CTI_Nano_PID) },
 	{ USB_DEVICE(FTDI_VID, FTDI_AMC232_PID) },
 	{ USB_DEVICE(FTDI_VID, FTDI_CANUSB_PID) },
 	{ USB_DEVICE(FTDI_VID, FTDI_CANDAPTER_PID) },
diff -Nru kernel-2.6.33mamba.orig//drivers/usb/serial/ftdi_sio_ids.h kernel-2.6.33mamba/drivers/usb/serial/ftdi_sio_ids.h
--- kernel-2.6.33mamba.orig//drivers/usb/serial/ftdi_sio_ids.h	2010-09-29 13:08:53.457119450 +0200
+++ kernel-2.6.33mamba/drivers/usb/serial/ftdi_sio_ids.h	2010-09-29 13:17:28.902429141 +0200
@@ -1085,6 +1085,12 @@
 #define XVERVE_SIGNALYZER_SH4_PID      0xBCA4
 
 /*
+ * CTI GmbH USB to RS485 converters (http://www.cti-lean.com)
+ */
+#define FTDI_CTI_Mini_PID      0xF608  /* USB-485-Mini der Firma CTI*/
+#define FTDI_CTI_Nano_PID      0xF60B  /* USB-Nano-485 def Firma CTI*/
+
+/*
  * Segway Robotic Mobility Platform USB interface (using VID 0x0403)
  * Submitted by John G. Rogers
  */
