diff -Nru kernel-2.6.19mamba.orig/drivers/video/vesafb-thread.c kernel-2.6.19mamba/drivers/video/vesafb-thread.c
--- kernel-2.6.19mamba.orig/drivers/video/vesafb-thread.c	2006-12-05 15:52:43.000000000 +0100
+++ kernel-2.6.19mamba/drivers/video/vesafb-thread.c	2006-12-05 15:55:24.000000000 +0100
@@ -16,6 +16,7 @@
 #include <linux/delay.h>
 #include <linux/signal.h>
 #include <linux/suspend.h>
+#include <linux/freezer.h>
 #include <linux/unistd.h>
 #include <video/vesa.h>
 #include <video/edid.h>
