--- linux-3.7.0-0.rc2.git4.2.fc19.x86_64/drivers/staging/omapdrm/omap_crtc.c.orig	2012-10-30 09:58:47.613641237 +0000
+++ linux-3.7.0-0.rc2.git4.2.fc19.x86_64/drivers/staging/omapdrm/omap_crtc.c	2012-10-30 10:05:36.996081758 +0000
@@ -19,9 +19,9 @@
 
 #include "omap_drv.h"
 
-#include "drm_mode.h"
-#include "drm_crtc.h"
-#include "drm_crtc_helper.h"
+#include <drm/drm_mode.h>
+#include <drm/drm_crtc.h>
+#include <drm/drm_crtc_helper.h>
 
 #define to_omap_crtc(x) container_of(x, struct omap_crtc, base)
 
