diff --git a/Makefile b/Makefile
index afb8e0d..e435bf5 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,5 @@
 VERSION = 3
 PATCHLEVEL = 0
-SUBLEVEL = 0
 EXTRAVERSION = -rc2
 NAME = Sneaky Weasel
 
@@ -378,9 +377,9 @@ KBUILD_LDFLAGS_MODULE := -T $(srctree)/scripts/module-common.lds
 
 # Read KERNELRELEASE from include/config/kernel.release (if it exists)
 KERNELRELEASE = $(shell cat include/config/kernel.release 2> /dev/null)
-KERNELVERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
+KERNELVERSION = $(VERSION).$(PATCHLEVEL)$(EXTRAVERSION)
 
-export VERSION PATCHLEVEL SUBLEVEL KERNELRELEASE KERNELVERSION
+export VERSION PATCHLEVEL KERNELRELEASE KERNELVERSION
 export ARCH SRCARCH CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC
 export CPP AR NM STRIP OBJCOPY OBJDUMP
 export MAKE AWK GENKSYMS INSTALLKERNEL PERL UTS_MACHINE
@@ -1005,8 +1004,8 @@ endef
 
 define filechk_version.h
 	(echo \#define LINUX_VERSION_CODE $(shell                             \
-	expr $(VERSION) \* 65536 + $(PATCHLEVEL) \* 256 + $(SUBLEVEL));     \
-	echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))';)
+	expr $(VERSION) \* 65536 + $(PATCHLEVEL) \* 256);     \
+	echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8))';)
 endef
 
 include/linux/version.h: $(srctree)/Makefile FORCE
