diff --git a/arch/arm/mach-exynos/mach-origen.c b/arch/arm/mach-exynos/mach-origen.c
index 4e574c2..5028fee 100644
--- a/arch/arm/mach-exynos/mach-origen.c
+++ b/arch/arm/mach-exynos/mach-origen.c
@@ -121,6 +121,7 @@ static struct regulator_consumer_supply __initdata ldo14_consumer[] = {
 };
 static struct regulator_consumer_supply __initdata ldo17_consumer[] = {
 	REGULATOR_SUPPLY("vdd33", "swb-a31"), /* AR6003 WLAN & CSR 8810 BT */
+	REGULATOR_SUPPLY("vmmc", NULL),
 };
 static struct regulator_consumer_supply __initdata buck1_consumer[] = {
 	REGULATOR_SUPPLY("vdd_arm", NULL), /* CPUFREQ */
