diff --color -u -r GE-Proton7-8/proton GE-Proton7-8-patched/proton
--- a/GE-Proton7-18/proton	2022-03-12 13:17:38.888543306 +0800
+++ b/GE-Proton7-18/proton	2022-03-12 13:18:46.559290671 +0800
@@ -734,6 +734,10 @@
 
             self.migrate_user_paths()
 
+            if not file_exists(self.prefix_dir + "dosdevices", follow_symlinks=False):
+                makedirs(self.prefix_dir + "dosdevices")
+                set_dir_casefold_bit(self.prefix_dir + "dosdevices")
+
             if not file_exists(self.prefix_dir + "/dosdevices/c:", follow_symlinks=False):
                 os.symlink("../drive_c", self.prefix_dir + "/dosdevices/c:")
 
