commit 2fedd8bd32de83b5d989e2e19910f200411f8224
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Tue May 31 11:51:38 2011 +0400

    Fix dovecot 2.0 conf.

diff --git a/profile/mail/etc/dovecot/dovecot.conf.2.0 b/profile/mail/etc/dovecot/dovecot.conf.2.0
index 99c4e15..cf13471 100644
--- a/profile/mail/etc/dovecot/dovecot.conf.2.0
+++ b/profile/mail/etc/dovecot/dovecot.conf.2.0
@@ -34,5 +34,6 @@ protocols = pop3
 }
 !ssl_cert_file = del
 !ssl_key_file = del
+!listen = del
 
 !include conf.d/*.conf
