commit a81b3ea474d1ca82c35f468768100205ad7e1eb1
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Fri May 20 17:30:59 2011 +0400

    Add /var/bind for named configuration.

diff --git a/profile/dns/etc/bind/named.conf b/profile/dns/etc/bind/named.conf
index b37b67f..53e5d4f 100644
--- a/profile/dns/etc/bind/named.conf
+++ b/profile/dns/etc/bind/named.conf
@@ -2,6 +2,8 @@
 chmod=0600\
 chown=named:named
 options {
+	// директория кэшей
+	directory "/var/bind";
 	// Доступные сети
 	#-sr_dns_net_allow_pass-#
 };
