commit b450c3dad3c7aa06d213342acca5fd15d0facca5
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Fri Jun 3 12:45:58 2011 +0400

    Change history-domain to history-domains.

diff --git a/i18n/cl_server_ru.mo b/i18n/cl_server_ru.mo
index 598fec3..77e1fc6 100644
Binary files a/i18n/cl_server_ru.mo and b/i18n/cl_server_ru.mo differ
diff --git a/pym/cl_ldap.py b/pym/cl_ldap.py
index c9d6279..e1c6eae 100644
--- a/pym/cl_ldap.py
+++ b/pym/cl_ldap.py
@@ -1237,16 +1237,16 @@ in a sambaDomainName',
     def getMailHistoryData(self,options):
         """Get mail history params from env and cmdline"""
         if not "history" in options and ("history-path" in options or
-                                         "history-domain" in options):
+                                         "history-domains" in options):
             if "history-path" in options:
                 param = "history-path"
             else:
-                param = "history-domain"
+                param = "history-domains"
             self.printERROR(_("Command line incorrect, "
                       "'%s' is not using without 'history'")%param)
             return False
-        if "history-domain" in options:
-            history_domain = options["history-domain"].strip()
+        if "history-domains" in options:
+            history_domain = options["history-domains"].strip()
             self.clVars.Set("sr_mail_history_domain", history_domain)
         else:
             history_domain = self.clVars.Get("sr_mail_history_domain") 
@@ -10103,9 +10103,9 @@ example: --range 192.168.0.2,192.168.0.50')
         },
         {'progAccess':(6,9),
          'optVal':"DOMAIN",
-         'longOption':"history-domain",
+         'longOption':"history-domains",
          'helpChapter':_("Mail service options"),
-         'help':_("domain for own mail")
+         'help':_("domain or domains for own mail (comma delimited)")
         },
         {'progAccess':(6,9),
          'optVal':"PATH",
