commit 2741abd56ac8c145646e5bbad5ccc243c4c1aa8a
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Thu Jul 14 13:26:29 2011 +0400

    Fix http://www.calculate-linux.ru/boards/33/topics/9827.

diff --git a/pym/cl_ldap.py b/pym/cl_ldap.py
index ebab6bd..6682a2d 100644
--- a/pym/cl_ldap.py
+++ b/pym/cl_ldap.py
@@ -16852,6 +16852,8 @@ class servDns(shareLdap):
         """
         hostName, spl, zoneName = domainName.partition(".")
         relZoneDN = self.getRelZoneDN(zoneName)
+        if hostName == "*":
+            hostName = r"\*"
         resSearch = self.searchLdapDN(hostName, relZoneDN, "relativeDomainName")
         return resSearch
 
