commit e125137ad2ac373bcf5816567d8209a9a1b11920
Author: Самоукин Алексей <as@calculate.ru>
Date:   Fri Apr 1 12:36:31 2011 +0400

    Rename cl-update to cl-update-server.

diff --git pym/cl_ldap.py pym/cl_ldap.py
index e76f252..6806945 100644
--- pym/cl_ldap.py
+++ pym/cl_ldap.py
@@ -8887,7 +8887,7 @@ class cl_ldap(shareLdap):
                           'cl-setup':6,
                           'cl-passwd':7,
                           'cl-backup':8,
-                          'cl-update':9,
+                          'cl-update-server':9,
                           'cl-rebuild':10,
                           'execserv':11,
                           'cl-replication':12,
@@ -11139,7 +11139,7 @@ service")
         {
          'progAccess':(9,),
          'helpChapter':_("Examples"),
-         'help':pcs("  cl-update samba", self.column_width,
+         'help':pcs("  cl-update-server samba", self.column_width,
              "# "+_("update config files Samba services") + ".",
              self.consolewidth-self.column_width)
         },
diff --git scripts/cl-update scripts/cl-update
deleted file mode 100644
index 752dcde..0000000
--- scripts/cl-update
+++ /dev/null
@@ -1,42 +0,0 @@
-#!/usr/bin/python
-#-*- coding: utf-8 -*-
-
-# Copyright 2008-2010 Mir Calculate Ltd. http://www.calculate-linux.org
-#
-#  Licensed under the Apache License, Version 2.0 (the "License");
-#  you may not use this file except in compliance with the License.
-#  You may obtain a copy of the License at
-#
-#      http://www.apache.org/licenses/LICENSE-2.0
-#
-#  Unless required by applicable law or agreed to in writing, software
-#  distributed under the License is distributed on an "AS IS" BASIS,
-#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-#  See the License for the specific language governing permissions and
-#  limitations under the License.
-
-import sys
-import os
-sys.path.insert(0,os.path.abspath('/usr/lib/calculate/calculate-lib/pym'))
-sys.path.insert(0,os.path.abspath('/usr/lib/calculate/calculate-server/pym'))
-import cl_base
-tr = cl_base.lang()
-tr.setGlobalDomain('cl_server')
-tr.setLanguage(sys.modules[__name__])
-import cl_ldap
-
-if __name__ == "__main__":
-    ldapObj = cl_ldap.cl_ldap("cl-update")
-    optObj = cl_ldap.tsOpt(ldapObj,[])
-    flagError = False
-    if not optObj.flagHelp and optObj.params.has_key('service') and\
-        not optObj.errorOpt:
-        # Настройка сервера LDAP
-        flagError = True
-        obj = cl_ldap.servLdap()
-        if obj.updateLdapServer(optObj.opt, optObj.params['service']):
-            flagError = False
-    if flagError:
-        sys.exit(1)
-    else:
-        sys.exit(0)
\ No newline at end of file
diff --git scripts/cl-update-server scripts/cl-update-server
new file mode 100644
index 0000000..d65a753
--- /dev/null
+++ scripts/cl-update-server
@@ -0,0 +1,42 @@
+#!/usr/bin/python
+#-*- coding: utf-8 -*-
+
+# Copyright 2008-2010 Mir Calculate Ltd. http://www.calculate-linux.org
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import sys
+import os
+sys.path.insert(0,os.path.abspath('/usr/lib/calculate/calculate-lib/pym'))
+sys.path.insert(0,os.path.abspath('/usr/lib/calculate/calculate-server/pym'))
+import cl_base
+tr = cl_base.lang()
+tr.setGlobalDomain('cl_server')
+tr.setLanguage(sys.modules[__name__])
+import cl_ldap
+
+if __name__ == "__main__":
+    ldapObj = cl_ldap.cl_ldap("cl-update-server")
+    optObj = cl_ldap.tsOpt(ldapObj,[])
+    flagError = False
+    if not optObj.flagHelp and optObj.params.has_key('service') and\
+        not optObj.errorOpt:
+        # Настройка сервера LDAP
+        flagError = True
+        obj = cl_ldap.servLdap()
+        if obj.updateLdapServer(optObj.opt, optObj.params['service']):
+            flagError = False
+    if flagError:
+        sys.exit(1)
+    else:
+        sys.exit(0)
\ No newline at end of file
diff --git setup.py setup.py
index 88c87e9..d0e34be 100755
--- setup.py
+++ setup.py
@@ -154,7 +154,7 @@ setup(
              "./scripts/cl-passwd",
              "./scripts/cl-setup",
              "./scripts/cl-backup",
-             "./scripts/cl-update",
+             "./scripts/cl-update-server",
              "./scripts/cl-rebuild",
              "./scripts/cl-replication",
              "./scripts/execserv",
