commit a75af744f9cd2e45404fad74ac560fabe1f1cc79
Author: Mike Hiretsky <mh@calculate.ru>
Date:   Tue May 17 10:05:20 2011 +0400

    Add processing fifo files.

diff --git a/pym/cl_ldap.py b/pym/cl_ldap.py
index 6806945..c66326b 100644
--- a/pym/cl_ldap.py
+++ b/pym/cl_ldap.py
@@ -1176,6 +1176,9 @@ in a sambaDomainName',
         for socketRm in scanObjs[0].sockets:
             # Удаляем сокеты
             os.remove(socketRm)
+        for fifoRm in scanObjs[0].fifo:
+            # Удаляем fifo
+            os.remove(fifoRm)
         for linkRm in scanObjs[0].links:
             # Удаляем ссылки
             os.unlink(linkRm[1])
@@ -1421,6 +1424,9 @@ if you want to continue to run the program again"))
             # Удаляем сокеты
             for rmSocket in scanObjs[0].sockets:
                 os.remove(rmSocket)
+            # Удаляем fifo
+            for rmFifo in scanObjs[0].fifo:
+                os.remove(rmFifo)
             mode,uid,gid = fileObj.getModeFile(srcDir)
             os.chmod(destDir, mode)
             os.chown(destDir, uid,gid)
