Index: supertux2.desktop
===================================================================
--- supertux2.desktop	(revision 5784)
+++ supertux2.desktop	(working copy)
@@ -35,8 +35,8 @@
 Comment[sv]=Ett klassiskt tvådimensionellt plattformsspel
 Comment[pt_BR]=Encarne o pinguin Tux neste jogo inspirado em clássicos de Pular&Correr
 Comment[hu]=Egy klasszikus 2D-s oldalnézeti játék
-Icon=supertux.png
+Icon=supertux2.png
 Exec=supertux2
 Terminal=false
 StartupNotify=false
-Categories=Game;ArcadeGame;
+Categories=Application;Game;ArcadeGame;
