--- a/bindings/CMakeLists.txt
+++ b/bindings/CMakeLists.txt
@@ -1,15 +1,5 @@
-find_package(Corrosion QUIET)
-if(NOT Corrosion_FOUND)
-    include(FetchContent)
-
-    FetchContent_Declare(
-        Corrosion
-        GIT_REPOSITORY https://github.com/corrosion-rs/corrosion.git
-        GIT_TAG v0.5.2
-    )
-
-    FetchContent_MakeAvailable(Corrosion)
-endif()
+# Use system-installed corrosion package
+find_package(Corrosion REQUIRED)
 
 corrosion_import_crate(
     MANIFEST_PATH Cargo.toml
