--- gnome-exe-thumbnailer.sh	2009-10-20 19:20:17.000000000 +0800
+++ gnome-exe-thumbnailer.sh	2010-05-05 16:46:06.299325202 +0800
@@ -21,17 +21,9 @@
 
 	icotool -x --index=$id $f -o $f
 
-	if [ $(file -b $f | cut -f3 -d' ') -gt 48 ]; then
-		convert -resize 48x48 $f $f
-	fi
-	
-	if [[ $3 == "--clean" ]]; then
-		cp $f $2
-	else
-		composite -gravity center $f /usr/share/pixmaps/gnome-exe-thumbnailer-template-wine.png $2
-	fi
+	cp $f $2
 else
-	cp /usr/share/pixmaps/gnome-exe-thumbnailer-template-wine.png "$2"
+	cp /usr/share/pixmaps/gnome-exe-thumbnailer-template.png "$2"
 fi
 
 rm $f
