diff -urN debian/config/config debian.fixed/config/config
--- debian/config/config	2019-05-05 10:59:14.000000000 -0400
+++ debian.fixed/config/config	2019-05-06 13:03:53.445432253 -0400
@@ -77,7 +77,7 @@
 #. Actually a file containing X.509 certificates, not keys.
 #. Whenever the filename changes, this also needs to be updated in
 #. debian/featureset-*/config
-CONFIG_SYSTEM_TRUSTED_KEYS="debian/certs/debian-uefi-certs.pem"
+CONFIG_SYSTEM_TRUSTED_KEYS=""
 #. Add secondary keyring with keys from UEFI db and MOK.
 CONFIG_SECONDARY_TRUSTED_KEYRING=y
 CONFIG_SYSTEM_BLACKLIST_KEYRING=y
