From bb4cba9b4dcdb339958232839dce68a646e441e3 Mon Sep 17 00:00:00 2001
From: Sly_tom_cat <slytomcat@mail.ru>
Date: Tue, 28 Nov 2017 17:35:54 +0300
Subject: [PATCH] Change initialization behavior (fix for #181)

---
 yandex-disk-indicator.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/yandex-disk-indicator.py b/yandex-disk-indicator.py
index be12ecf..2589b5a 100644
--- a/yandex-disk-indicator.py
+++ b/yandex-disk-indicator.py
@@ -912,7 +912,7 @@ def update(self, vals, update, yddir):  # Update information in menu
         self.used.set_label(_('Used: ') + vals['used'] + '/' + vals['total'])
         self.free.set_label(_('Free: ') + vals['free'] + _(', trash: ') + vals['trash'])
       # Update last synchronized sub-menu when daemon is running
-      if (update.last or update.init) and vals['status'] != 'none':
+      if (update.last or update.init):                # Unnecessary: and vals['status'] != 'none':
         for widget in self.lastItems.get_children():  # Clear last synchronized sub-menu
           self.lastItems.remove(widget)
         for filePath in vals['lastitems']:            # Create new sub-menu items
