diff -ur a/yandex-disk-indicator.py b/yandex-disk-indicator.py
--- a/yandex-disk-indicator.py	2018-10-22 17:23:11.000000000 +0300
+++ b/yandex-disk-indicator.py	2018-10-31 18:49:11.443968583 +0300
@@ -781,6 +781,11 @@
       # Update last synchronized sub-menu on first run or when last data has changed
       if vals['lastchg'] or vals['laststatus'] == 'unknown':
         # Update last synchronized sub-menu
+        
+        # Fix bug: https://github.com/slytomcat/yandex-disk-indicator/issues/197
+        # Temporary disable show synchronized sub menu while updating
+        self.lastItems.destroy()
+        
         self.lastItems = Gtk.Menu()                   # Create new Sub-menu:
         for filePath in vals['lastitems']:            # Create new sub-menu items
           # Create menu label as file path (shorten it down to 50 symbols when path length > 50
