--- linux-2.6.22-rc3-orig/drivers/net/usb/rndis_host.c	2007-05-25 22:55:14.000000000 -0400
+++ linux-2.6.22-rc3/drivers/net/usb/rndis_host.c	2007-05-27 17:06:16.000000000 -0400
@@ -499,8 +499,7 @@
 	net->hard_header_len += sizeof (struct rndis_data_hdr);
 	dev->hard_mtu = net->mtu + net->hard_header_len;
 
-	dev->rx_urb_size = dev->hard_mtu + (dev->maxpacket + 1);
-	dev->rx_urb_size &= ~(dev->maxpacket - 1);
+	dev->rx_urb_size = (dev->udev->speed == USB_SPEED_FULL) ? 16384 : 8192;
 	u.init->max_transfer_size = cpu_to_le32(dev->rx_urb_size);
 
 	net->change_mtu = NULL;
