diff --git a/setup/Linux/oss/build/osscore.c b/setup/Linux/oss/build/osscore.c
index d79c6a6..1cc89fc 100644
--- a/setup/Linux/oss/build/osscore.c
+++ b/setup/Linux/oss/build/osscore.c
@@ -472,11 +472,14 @@ oss_get_pid (void)
 int
 oss_get_uid (void)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,29)
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(3,12,0)
+  kuid_t uid = current_uid();
+  return uid.val;
+ #elif LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,29)
   return current->cred->uid;
-#else
+ #else 	  	
   return current->uid;
-#endif
+ #endif
 }
 
 typedef struct tmout_desc
@@ -991,6 +994,8 @@ alloc_fop (oss_device_t * osdev, struct oss_file_operation_handle *op)
   fop->write = tmp_write;
 #if LINUX_VERSION_CODE < KERNEL_VERSION(3,11,0)
   fop->readdir = NULL;		/* tmp_readdir; */
+#else
+  fop->iterate = NULL;
 #endif
   fop->poll = tmp_poll;
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,35)
