--- src/org.mate.panel.DockApplet.mate-panel-applet.in.orig	2015-11-07 11:05:00.238396421 +0200
+++ src/org.mate.panel.DockApplet.mate-panel-applet.in	2015-11-07 11:06:24.123403521 +0200
@@ -1,7 +1,7 @@
 [Applet Factory]
 Id=DockAppletFactory
 InProcess=false
-Location=@LOCATION@
+Location=/usr/bin/python3 /usr/lib64/mate-applets/mate-dock-applet/dock_applet.py
 Name=Dock Applet Factory
 Description=An application dock for the MATE panel
 
--- src/org.mate.panel.applet.DockAppletFactory.service.in.orig	2015-11-07 11:05:30.078398947 +0200
+++ src/org.mate.panel.applet.DockAppletFactory.service.in	2015-11-07 11:07:14.003407743 +0200
@@ -1,3 +1,3 @@
 [D-BUS Service]
 Name=org.mate.panel.applet.DockAppletFactory
-Exec=/usr/bin/env python3  @LOCATION@
+Exec=/usr/bin/python3 /usr/lib64/mate-applets/mate-dock-applet/dock_applet.py
