--- a/kernel/workqueue.c	2012-12-11 16:04:29.683383186 +0200
+++ b/kernel/workqueue.c	2012-12-11 16:04:54.516089824 +0200
@@ -1350,7 +1350,7 @@
 	/* should have been called from irqsafe timer with irq already off */
 	__queue_work(dwork->cpu, cwq->wq, &dwork->work);
 }
-EXPORT_SYMBOL_GPL(delayed_work_timer_fn);
+EXPORT_SYMBOL(delayed_work_timer_fn);
 
 static void __queue_delayed_work(int cpu, struct workqueue_struct *wq,
 				struct delayed_work *dwork, unsigned long delay)
