--- xgamma-gui-9999/Makefile.orig	2014-12-17 19:09:53.298076090 +0200
+++ xgamma-gui-9999/Makefile	2014-12-17 19:10:14.215445931 +0200
@@ -21,7 +21,7 @@
 LIBS          = $(SUBLIBS) -L/usr/X11R6/lib64 -lQt5Widgets -L/usr/lib/x86_64-linux-gnu -lQt5Gui -lQt5Core -lGL -lpthread 
 AR            = ar cqs
 RANLIB        = 
-QMAKE         = /usr/lib/x86_64-linux-gnu/qt5/bin/qmake
+QMAKE         = /usr/lib64/qt5/bin/qmake
 TAR           = tar -cf
 COMPRESS      = gzip -9f
 COPY          = cp -f
