From: Mike Hommey <mh@glandium.org>
Date: Sat, 18 Nov 2017 18:34:52 +0900
Subject: Don't err on tildes when using shellutil.split

---
 python/mozbuild/mozbuild/shellutil.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/python/mozbuild/mozbuild/shellutil.py b/python/mozbuild/mozbuild/shellutil.py
index 185a970ee533..580d778739b0 100644
--- a/python/mozbuild/mozbuild/shellutil.py
+++ b/python/mozbuild/mozbuild/shellutil.py
@@ -26,7 +26,7 @@ UNQUOTED_TOKENS_RE = _tokens2re(
   whitespace=r'[\t\r\n ]+',
   quote=r'[\'"]',
   comment='#',
-  special=r'[<>&|`~(){}$;\*\?]',
+  special=r'[<>&|`(){}$;\*\?]',
   backslashed=r'\\[^\\]',
 )
 
