diff -upNr linux-3.5.0.orig//fs/zfs/Makefile linux-3.5.0/fs/zfs/Makefile
--- linux-3.5.0.orig//fs/zfs/Makefile	2012-07-29 10:40:22.491137000 -0400
+++ linux-3.5.0/fs/zfs/Makefile	2012-07-29 17:44:16.538156777 -0400
@@ -31,10 +31,10 @@ modules:
 clean:
 	@# Only cleanup the kernel build directories when CONFIG_KERNEL
 	@# is defined.  This indicates that kernel modules should be built.
-	$(MAKE) -C /lib/modules/3.5.0/build SUBDIRS=`pwd`  O=/lib/modules/3.5.0/build $@
+#	$(MAKE) -C /lib/modules/3.5.0/build SUBDIRS=`pwd`  O=/lib/modules/3.5.0/build $@
 
 	if [ -f Module.symvers ]; then $(RM) Module.symvers; fi
-	if [ -f Module.symvers ]; then $(RM) Module.symvers; fi
+	if [ -f NONE ]; then $(RM) NONE; fi
 	if [ -f Module.markers ]; then $(RM) Module.markers; fi
 
 modules_install:
diff -upNr linux-3.5.0.orig//include/spl/Makefile linux-3.5.0/include/spl/Makefile
--- linux-3.5.0.orig//include/spl/Makefile	2012-07-29 10:36:20.678797000 -0400
+++ linux-3.5.0/include/spl/Makefile	2012-07-29 17:44:16.592156775 -0400
@@ -119,7 +119,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LICENSE = GPL
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
diff -upNr linux-3.5.0.orig//include/zfs/linux/Makefile linux-3.5.0/include/zfs/linux/Makefile
--- linux-3.5.0.orig//include/zfs/linux/Makefile	2012-07-29 10:40:22.456222000 -0400
+++ linux-3.5.0/include/zfs/linux/Makefile	2012-07-29 17:44:16.607156775 -0400
@@ -215,7 +215,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -340,8 +340,8 @@ USER_H =
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs/linux
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/linux
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/linux
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-am
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//include/zfs/Makefile linux-3.5.0/include/zfs/Makefile
--- linux-3.5.0.orig//include/zfs/Makefile	2012-07-29 10:40:22.469974000 -0400
+++ linux-3.5.0/include/zfs/Makefile	2012-07-29 17:44:16.608156775 -0400
@@ -268,7 +268,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -405,8 +405,8 @@ USER_H = \
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-recursive
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//include/zfs/sys/fm/fs/Makefile linux-3.5.0/include/zfs/sys/fm/fs/Makefile
--- linux-3.5.0.orig//include/zfs/sys/fm/fs/Makefile	2012-07-29 10:40:22.467974000 -0400
+++ linux-3.5.0/include/zfs/sys/fm/fs/Makefile	2012-07-29 17:44:16.613156775 -0400
@@ -213,7 +213,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -335,8 +335,8 @@ USER_H =
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs/sys/fm/fs
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fm/fs
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fm/fs
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-am
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//include/zfs/sys/fm/Makefile linux-3.5.0/include/zfs/sys/fm/Makefile
--- linux-3.5.0.orig//include/zfs/sys/fm/Makefile	2012-07-29 10:40:22.468847000 -0400
+++ linux-3.5.0/include/zfs/sys/fm/Makefile	2012-07-29 17:44:16.614156775 -0400
@@ -253,7 +253,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -377,8 +377,8 @@ USER_H =
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs/sys/fm
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fm
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fm
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-recursive
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//include/zfs/sys/fs/Makefile linux-3.5.0/include/zfs/sys/fs/Makefile
--- linux-3.5.0.orig//include/zfs/sys/fs/Makefile	2012-07-29 10:40:22.467009000 -0400
+++ linux-3.5.0/include/zfs/sys/fs/Makefile	2012-07-29 17:44:16.615156775 -0400
@@ -213,7 +213,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -335,8 +335,8 @@ USER_H =
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs/sys/fs
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fs
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys/fs
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-am
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//include/zfs/sys/Makefile linux-3.5.0/include/zfs/sys/Makefile
--- linux-3.5.0.orig//include/zfs/sys/Makefile	2012-07-29 10:40:22.465801000 -0400
+++ linux-3.5.0/include/zfs/sys/Makefile	2012-07-29 17:44:16.616156775 -0400
@@ -395,7 +395,7 @@ LIBTOOL = $(SHELL) $(top_builddir)/libto
 LIBUUID = -luuid
 LINUX = /usr/src/linux
 LINUX_OBJ = /lib/modules/3.5.0/build
-LINUX_SYMBOLS = Module.symvers
+LINUX_SYMBOLS = NONE
 LINUX_VERSION = 3.5.0
 LIPO = 
 LN_S = ln -s
@@ -593,8 +593,8 @@ USER_H =
 EXTRA_DIST = $(COMMON_H) $(KERNEL_H) $(USER_H)
 libzfsdir = $(includedir)/libzfs/sys
 libzfs_HEADERS = $(COMMON_H) $(USER_H)
-kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys
-kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
+#kerneldir = /usr/src/zfs-$(ZFS_META_VERSION)-$(ZFS_META_RELEASE)/$(LINUX_VERSION)/sys
+#kernel_HEADERS = $(COMMON_H) $(KERNEL_H)
 all: all-recursive
 
 .SUFFIXES:
diff -upNr linux-3.5.0.orig//spl/Makefile linux-3.5.0/spl/Makefile
--- linux-3.5.0.orig//spl/Makefile	2012-07-29 10:36:20.692439000 -0400
+++ linux-3.5.0/spl/Makefile	2012-07-29 17:44:16.649156776 -0400
@@ -13,9 +13,9 @@ modules:
 clean:
 	@# Only cleanup the kernel build directories when CONFIG_KERNEL
 	@# is defined.  This indicates that kernel modules should be built.
-	$(MAKE) -C /lib/modules/3.5.0/build SUBDIRS=`pwd`  O=/lib/modules/3.5.0/build $@
+#	$(MAKE) -C /lib/modules/3.5.0/build SUBDIRS=`pwd`  O=/lib/modules/3.5.0/build $@
 
-	if [ -f 'Module.symvers' ]; then $(RM) 'Module.symvers'; fi
+	if [ -f 'NONE' ]; then $(RM) 'NONE'; fi
 	if [ -f Module.markers ]; then $(RM) Module.markers; fi
 
 modules_install:
diff -upNr linux-3.5.0.orig//spl_config.h linux-3.5.0/spl_config.h
--- linux-3.5.0.orig//spl_config.h	2012-07-29 10:36:20.697489000 -0400
+++ linux-3.5.0/spl_config.h	2012-07-29 17:44:16.660156777 -0400
@@ -86,7 +86,7 @@
 #define HAVE_FILES_FDTABLE 1
 
 /* first_online_pgdat() is available */
-/* #undef HAVE_FIRST_ONLINE_PGDAT */
+#define HAVE_FIRST_ONLINE_PGDAT 1
 
 /* fls64() is available */
 #define HAVE_FLS64 1
@@ -104,7 +104,7 @@
 #define HAVE_GLOBAL_PAGE_STATE 1
 
 /* groups_search() is available */
-/* #undef HAVE_GROUPS_SEARCH */
+#define HAVE_GROUPS_SEARCH 1
 
 /* init_utsname() is available */
 #define HAVE_INIT_UTSNAME 1
@@ -128,7 +128,7 @@
 #define HAVE_KERN_PATH_PARENT_HEADER 1
 
 /* kern_path_parent() is available */
-/* #undef HAVE_KERN_PATH_PARENT_SYMBOL */
+#define HAVE_KERN_PATH_PARENT_SYMBOL 1
 
 /* kmalloc_node() is available */
 #define HAVE_KMALLOC_NODE 1
@@ -152,10 +152,10 @@
 #define HAVE_MUTEX_OWNER_TASK_STRUCT 1
 
 /* next_online_pgdat() is available */
-/* #undef HAVE_NEXT_ONLINE_PGDAT */
+#define HAVE_NEXT_ONLINE_PGDAT 1
 
 /* next_zone() is available */
-/* #undef HAVE_NEXT_ZONE */
+#define HAVE_NEXT_ZONE 1
 
 /* struct path used in struct nameidata */
 #define HAVE_PATH_IN_NAMEIDATA 1
@@ -173,7 +173,7 @@
 #define HAVE_PUT_TASK_STRUCT 1
 
 /* set_fs_pwd() is available */
-/* #undef HAVE_SET_FS_PWD */
+#define HAVE_SET_FS_PWD 1
 
 /* set_normalized_timespec() is available as export */
 #define HAVE_SET_NORMALIZED_TIMESPEC_EXPORT 1
@@ -212,7 +212,7 @@
 #define HAVE_SYS_TYPES_H 1
 
 /* task_curr() is available */
-/* #undef HAVE_TASK_CURR */
+#define HAVE_TASK_CURR 1
 
 /* timespec_sub() is available */
 #define HAVE_TIMESPEC_SUB 1
diff -upNr linux-3.5.0.orig//zfs_config.h linux-3.5.0/zfs_config.h
--- linux-3.5.0.orig//zfs_config.h	2012-07-29 10:40:22.493751000 -0400
+++ linux-3.5.0/zfs_config.h	2012-07-29 17:44:16.660156777 -0400
@@ -71,7 +71,7 @@
 #define HAVE_BLK_QUEUE_FLUSH 1
 
 /* blk_queue_flush() is GPL-only */
-#define HAVE_BLK_QUEUE_FLUSH_GPL_ONLY 1
+/* #undef HAVE_BLK_QUEUE_FLUSH_GPL_ONLY */
 
 /* blk_queue_io_opt() is available */
 #define HAVE_BLK_QUEUE_IO_OPT 1