diff -x .git -urN linux-3.9/net/ipv4/Kconfig zen/net/ipv4/Kconfig
--- linux-3.9/net/ipv4/Kconfig	2013-04-29 03:36:01.000000000 +0300
+++ zen/net/ipv4/Kconfig	2013-05-01 11:55:03.956679287 +0300
@@ -592,6 +592,9 @@
 	config DEFAULT_VEGAS
 		bool "Vegas" if TCP_CONG_VEGAS=y
 
+	config DEFAULT_YEAH
+		bool "YeAH" if TCP_CONG_YEAH=y
+
 	config DEFAULT_VENO
 		bool "Veno" if TCP_CONG_VENO=y
 
@@ -617,6 +620,7 @@
 	default "htcp" if DEFAULT_HTCP
 	default "hybla" if DEFAULT_HYBLA
 	default "vegas" if DEFAULT_VEGAS
+	default "yeah" if DEFAULT_YEAH
 	default "westwood" if DEFAULT_WESTWOOD
 	default "veno" if DEFAULT_VENO
 	default "reno" if DEFAULT_RENO
