--- linux-3.8.0-0.rc2.git1.1.fc19.x86_64/drivers/staging/nvec/Kconfig.orig	2013-01-07 11:04:43.493510550 +0000
+++ linux-3.8.0-0.rc2.git1.1.fc19.x86_64/drivers/staging/nvec/Kconfig	2013-01-07 11:14:18.186033211 +0000
@@ -1,6 +1,6 @@
 config MFD_NVEC
 	bool "NV Tegra Embedded Controller SMBus Interface"
-	depends on I2C && GPIOLIB && ARCH_TEGRA
+	depends on I2C && GPIOLIB && ARCH_TEGRA && MFD_CORE
 	select MFD_CORE
 	help
 	    Say Y here to enable support for a nVidia compliant embedded
