diff -urN linux-3.7/scripts/mkcompile_h zen-zen-tag-12095/scripts/mkcompile_h
--- linux-3.7/scripts/mkcompile_h	2012-12-11 05:30:57.000000000 +0200
+++ zen-zen-tag-12095/scripts/mkcompile_h	2013-01-21 15:55:35.173825052 +0200
@@ -54,8 +54,8 @@
 fi
 
 UTS_VERSION="#$VERSION"
-CONFIG_FLAGS=""
-if [ -n "$SMP" ] ; then CONFIG_FLAGS="SMP"; fi
+CONFIG_FLAGS="ZEN"
+if [ -n "$SMP" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS SMP"; fi
 if [ -n "$PREEMPT" ] ; then CONFIG_FLAGS="$CONFIG_FLAGS PREEMPT"; fi
 UTS_VERSION="$UTS_VERSION $CONFIG_FLAGS $TIMESTAMP"
 
