--- linux-3.3.4-3.fc17.x86_64_orig/drivers/usb/core/hub.c	2012-05-02 20:08:18.421685932 -0400
+++ linux-3.3.4-3.fc17.x86_64/drivers/usb/core/hub.c	2012-05-02 20:30:36.565865425 -0400
@@ -3484,6 +3484,13 @@ static void hub_events(void)
 				(u16) hub->change_bits[0],
 				(u16) hub->event_bits[0]);
 
+		/* Don't disconnect USB-SATA on TrimSlice */
+		if (strcmp(dev_name(hdev->bus->controller), "tegra-ehci.0") == 0) {
+			if ((hdev->state == 7) && (hub->change_bits[0] == 0) &&
+			    (hub->event_bits[0] == 0x2))
+				hub->event_bits[0] = 0;
+		}
+
 		/* Lock the device, then check to see if we were
 		 * disconnected while waiting for the lock to succeed. */
 		usb_lock_device(hdev);
