--- linux-3.3.0-4.fc17.noarch/kernel/time/timekeeping.c~	2012-03-30 14:18:15.591162207 -0400
+++ linux-3.3.0-4.fc17.noarch/kernel/time/timekeeping.c	2012-03-30 14:18:38.959121171 -0400
@@ -854,13 +854,6 @@ static void timekeeping_adjust(s64 offse
 	} else /* No adjustment needed */
 		return;
 
-	WARN_ONCE(timekeeper.clock->maxadj &&
-			(timekeeper.mult + adj > timekeeper.clock->mult +
-						timekeeper.clock->maxadj),
-			"Adjusting %s more then 11%% (%ld vs %ld)\n",
-			timekeeper.clock->name, (long)timekeeper.mult + adj,
-			(long)timekeeper.clock->mult +
-				timekeeper.clock->maxadj);
 	/*
 	 * So the following can be confusing.
 	 *
